### ISSUE TITLE
MSIECF: decode exception in HTTP headers output

### ISSUE THREAD

**Issue by joachimmetz** [log2timeline/plaso#287]
```
$ psort.py -o L2tcsv analyst018.db -w l2t_csv
[INFO] Data files will be loaded from /usr/share/plaso by default.
Traceback (most recent call last):
  File "/usr/bin/psort.py", line 646, in <module>
    if not Main():
  File "/usr/bin/psort.py", line 641, in Main
    tool.ProcessStorage()
  File "/usr/bin/psort.py", line 593, in ProcessStorage
    self._ProcessStorage()
  File "/usr/bin/psort.py", line 236, in _ProcessStorage
    time_slice=time_slice, use_time_slicer=self._use_time_slicer)
  File "/usr/lib/python2.7/dist-packages/plaso/frontend/psort.py", line 422, in ProcessStorage
    analysis_queues=event_queue_producers)
  File "/usr/lib/python2.7/dist-packages/plaso/output/interface.py", line 241, in __exit__
    self.End()
  File "/usr/lib/python2.7/dist-packages/plaso/output/interface.py", line 233, in End
    self.Flush()
  File "/usr/lib/python2.7/dist-packages/plaso/output/interface.py", line 198, in Flush
    self._output_module.WriteEvent(event_object)
  File "/usr/lib/python2.7/dist-packages/plaso/output/interface.py", line 55, in WriteEvent
    self.WriteEventBody(event_object)
  File "/usr/lib/python2.7/dist-packages/plaso/output/l2t_csv.py", line 77, in WriteEventBody
    event_object)
  File "/usr/lib/python2.7/dist-packages/plaso/output/mediator.py", line 122, in GetFormattedMessages
    return event_formatter.GetMessages(self._formatter_mediator, event_object)
  File "/usr/lib/python2.7/dist-packages/plaso/formatters/msiecf.py", line 34, in GetMessages
    event_values[u'http_headers'] = http_headers.replace(u'\r\n', u' - ')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xde in position 115: ordinal not in range(128)
```

```
$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import dfvfs
>>> print(dfvfs.__version__)
20150708
```

plaso - log2timeline version 1.3.0

Please let me know what other info I can provide to help debug this.

**Comment by joachimmetz** [comment]
Can you try to pin point this to the corresponding index.dat file.

What I suspect is that the [HTTP headers](https://github.com/libyal/libmsiecf/blob/master/documentation/MSIE%20Cache%20File%20%28index.dat%29%20format.asciidoc#441-cache-url-record-data) are not stored in ASCII but with a codepage. I would like to confirm this against the data.

**Comment by apapadop** [comment]
Certainly, if you can talk me through it. How do I pinpoint which index.dat file is read at this point? Notice this is psort.py reading an existing db file.

I tried adding --debug which gave me:

$ psort.py --debug -o L2tcsv analyst018.db -w l2t_csv
[INFO] Data files will be loaded from /usr/share/plaso by default.
[ERROR] 'ascii' codec can't decode byte 0xde in position 115: ordinal not in range(128)
> /usr/lib/python2.7/dist-packages/plaso/formatters/msiecf.py(34)GetMessages()
-> event_values[u'http_headers'] = http_headers.replace(u'\r\n', u' - ')
(Pdb) debug
ENTERING RECURSIVE DEBUGGER
> <string>(1)<module>()
((Pdb)) list
[EOF]
((Pdb)) args
((Pdb)) w
  /usr/lib/python2.7/bdb.py(400)run()
-> exec cmd in globals, locals
> <string>(1)<module>()
((Pdb)) whatis
*** SyntaxError: SyntaxError('unexpected EOF while parsing', ('<string>', 0, 0, ''))
((Pdb)) where
  /usr/lib/python2.7/bdb.py(400)run()
-> exec cmd in globals, locals
> <string>(1)<module>()
((Pdb))

**Comment by joachimmetz** [comment]
`(Pdb)`

When you are at this point see what the output is of:
```
p event_object.pathspec.comparable
```

**Comment by apapadop** [comment]
This works - I now have a pointer to the index.dat file. Can't share it publicly though - do you want me to run some tests on it, or send it you directly via some mechanism?

**Comment by apapadop** [comment]
Actually, here is the full output: (hint - this index.dat is in a VSS volume)
(Pdb) p event_object.pathspec.comparable
u'type: OS, location: /media/<snip>.E01\ntype: EWF\ntype: TSK_PARTITION, location: /p2, part index: 3, start offset: 0x06500000\ntype: VSHADOW, location: /vss7, store index: 6\ntype: TSK, inode: 125238, location: /Users/<snip>/AppData/Local/Microsoft/Windows/Temporary Internet Files/Content.IE5/index.dat\n'

**Comment by joachimmetz** [comment]
> Can't share it publicly though - do you want me to run some tests on it, or send it you directly via some mechanism?

If you can share it via mail that is the easiest otherwise I'll ask you to run msiecfexport on it with [debug and verbose output](https://github.com/libyal/libmsiecf/wiki/Troubleshooting#verbose-and-debug-output) and sending me the offending entry.

**Comment by apapadop** [comment]
Sent by email.

**Comment by joachimmetz** [comment]
From what I currently can make up is that might be an issue with handling trailing data in the HTTP header data. I'll have a closer look.

**Comment by kiddinn** [comment]
We had another report come in for this bug.

Asked for pdb debug (slightly sanitized):

```
(Pdb) event_values['http_headers']
'HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nX-SharePointHealthScore: 0\r\nX-AspNet-Version: XXXr\nSPRequestGuid: YYYY\r\nrequest-id: ZZZZ\r\nSPRequestDuration: 714\r\nSPIisLatency: 0\r\nPersistent-Auth: true\r\nX-Powered-By: ASP.NET\r\nMicrosoftSharePointTeamServices: XXX\r\nX-Content-Type-Options: nosniff\r\nX-MS-InvokeApp: 1; RequireReadOnly\r\nContent-Length: 10231URL \x06\x00\x00\x00\x80...<other hex code follows>'
```
The URL there is not unicode.

**Comment by kiddinn** [comment]
Other reports:

```
(Pdb) event_values['http_headers']
'HTTP/1.1 200 OK\r\nContent-Type: image/jpeg\r\nContent-Length: 29634URL \x04\x00\x00\x00\x00\xac\<REDACTED_HEX>k~\xcf\x01`^\xfea\<REDACTED_HEX>f\x01\x07E\xf1\x9c\x00\x00\x00\x00\xaa\xf6\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00h\x00'
```

Another one:

```
'HTTP/1.1 200 OK\r\nETag: <...> Content-Length: 1024\r\n\x8aCache-Control: no-cache, no-store\r\nContent-Type: application/javascript\r\n<...>
```

**Comment by ant1** [comment]
I got a similar backtrace
```
[ERROR] 'ascii' codec can't decode byte 0xe9 in position 14: ordinal not in range(128)
> /usr/lib/python2.7/dist-packages/plaso/formatters/msiecf.py(34)GetMessages()
-> event_values[u'http_headers'] = http_headers.replace(u'\r\n', u' - ')
(Pdb) print repr(http_headers)
'HTTP/1.1 302 D\xe9plac\xe9 Temporairement\r\nLocation: https://REDACTED\r\nContent-Type: text/html;charset=UTF-8\r\nContent-Length: 0\r\n\r\n~U:REDACTED\r\n'
```

Déplacé Temporairement = Moved Temporarily

**Comment by msoria** [comment]
Temporary patched line 33-34 with this but I do not think this should be the final solution:

    if http_headers:
-->      http_headers= http_headers.decode('utf-8','ignore')
      event_values[u'http_headers'] = http_headers.replace(u'\r\n', u' - ')

**Comment by smclinden** [comment]
Here is another example:

(Pdb) p event_object.pathspec.comparable
u'type: OS, location: J:\\REDACTED.img\ntype: RAW\ntype: TSK_PARTITION, location: /p1, part index: 2, start offset: 0x00007e00\ntype: TSK, inode: 12,594, location: /Documents and Settings/Administrator/Local Settings/Temporary Internet Files/Content.IE5/index.dat\n'

(Pdb) print repr(http_headers)
'HTTP/1.1 200 OK\r\nContent-Length: 0\r\nContent-Type: Je-wZ\\\xc0\x04\r\n\r'
(Pdb) print event_values[u'http_headers']
HTTP/1.1 200 OK
Content-Length: 0
Content-Type: Je-wZ\└♦

(Pdb) print event_values['http_headers']
HTTP/1.1 200 OK
Content-Length: 0
Content-Type: Je-wZ\└♦

(Pdb) print event_values
{'store_number': 6, 'parser': u'msiecf', 'cache_directory_index': 3, 'tag': None, 'inode': 12594, 'display_name': u'TSK:/Documents and Settings/Administrator/Local Settings/Temporary Internet Files/Content.IE5/index.dat', 'uuid': u'd4a7f7370b5141468b1ca852762fad7b', 'http_headers': 'HTTP/1.1 200 OK\r\nContent-Length: 0
\r\nContent-Type: Je-wZ\\\xc0\x04\r\n\r', 'number_of_hits': 1, 'hostname': u'XXXX', 'filename': u'/Documents and Settings/Administrator/Local Settings/Temporary Internet Files/Content.IE5/index.dat', 'pathspec': <dfvfs.path.tsk_path_spec.TSKPathSpec object at 0x0000000004F1BFD0>, 'recovered': False, 'username': u'-', 'cached_file_size': 0, 'data_type': u'msiecf:url', 'timestamp': 1267928605109000L
, 'store_index': 3536, 'cache_directory_name': 'XA18O3WP', 'offset': 77056, 'cached_filename': u'BPSAjax[1].aspx', 'timestamp_desc': u'Last Access Time', 'url': u'http://support.dell.com/Support/BPSAjax.aspx?page=&eivalue=BPS_Tool=ProductModel&_='}

**Comment by joachimmetz** [comment]
```
HTTP/1.1 302 D\xe9plac\xe9 Temporairement
```
Thanks for confirming the format actually stores the data encoded, now to figure out which encoding (system or iso-8859-1 or other)

```
29634URL 
```
@kiddinn this indicates an URL record detection issue - can you provide me more details.

```
Content-Type: Je-wZ\\xc0\x04\r\n\r
```
This one eludes me at the moment, any indication what this content type should be?

**Comment by joachimmetz** [comment]
Issue not fully solved, but should no longer be a 1.4 show stopper.
